include/native/stream_filter.h

```diff
@@ -22,128 +22,140 @@
-              //If is an end of stream element
-              if( !item.first && item.second== -1 ){
-                  nend++;
-                  if(nend == p.num_threads -1 ) break;
-              }
-              //If there is not an end element
-              else {
-                  //If the element is the next one to be procesed
-                  if(order == item.second){
-                      if(item.first)
-                         cons(item.first.value());
-                      order++;
-                  }else{
-                      //If the incoming element is disordered
-                      aux_vector.push_back(item);
-                  }
-              }
-              //Search in the vector for next elements
-              for(auto it = aux_vector.begin(); it < aux_vector.end();it++) {
-                  if((*it).second == order){
-                       if((*it).first)
-                          cons((*it).first.value());
-                       aux_vector.erase(it);
-                       order++;
-                  }
-              } 
-              item = outqueue.pop();
-           }
-           while(aux_vector.size()>0){
-               for(auto it = aux_vector.begin(); it < aux_vector.end();it++) {
-                  if((*it).second == order){
-                       if((*it).first)
-                          cons((*it).first.value());
-                       aux_vector.erase(it);
-                       order++;
-                  }
-              }
-           }
-           
-           // Deregister the thread in the execution model
-           p.deregister_thread();
+/**
+\brief Invoke [stream filter pattern](@ref md_stream-filter pattern) on a data
+sequence with sequential execution policy.
+\tparam Generator Callable type for value generator.
+\tparam Predicate Callable type for filter predicate.
+\tparam Consumer Callable type for value consumer.
+\param ex Native parallel execution policy object.
+\param generate_op Generator callable object.
+\param predicate_op Predicate callable object.
+\param consume_op Consumer callable object.
+*/
+template <typename Generator, typename Predicate, typename Consumer>
+void stream_filter(parallel_execution_native & ex, Generator generate_op, 
+                   Predicate predicate_op, Consumer consume_op) 
+{
+  using namespace std;
+  using generated_type = typename result_of<Generator()>::type;
+  using item_type = pair<generated_type,long>;
+
+  mpmc_queue<item_type> generated_queue{ex.queue_size,ex.lockfree};
+  mpmc_queue<item_type> filtered_queue{ex.queue_size,ex.lockfree};
+
+  //THREAD 1-(N-1) EXECUTE FILTER AND PUSH THE VALUE IF TRUE
+  vector<thread> tasks;
+  for (int i=0; i<ex.num_threads-1; ++i) {
+    tasks.emplace_back([&](){
+      ex.register_thread();
+
+      // queue a pair element - order
+      auto item{generated_queue.pop()};
+      while (item.first) {
+        if(predicate_op(item.first.value())) {
```

Comment:
You are right, although parent are superfluous:
~~~c++
if (predicate_op(*item.first))
~~~